Add torch._library.impl_abstract_class to fakify torchBind class

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122105
* #121155
* #121050
* __->__ #120045

We fakify the script object arguments of custom ops right before fake tensor mode dispatch. In order for the subsequent invocation of other custom ops on the same script object to get the same fakified script object, we cache the fakified object. 

It makes use of the `__hash__` method instead of directly key on the script object itself because script object doesn't support `__eq__` by default. This is also consistent with prior art in [here](https://github.com/pytorch/pytorch/blob/main/torch/_export/passes/lift_constants_pass.py#L26-L31) and here 


cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

Differential Revision: [D54136016](https://our.internmc.facebook.com/intern/diff/D54136016)

ydwu4 committed:
Add torch.library.impl_abstract_class

[ghstack-poisoned]

ydwu4 commented on torch/library.py:
```diff
@@ -473,6 +473,25 @@ def inner(func):
 
+registered_class: Dict[str, Any] = {}
+
+def impl_abstract_class(qualname, fake_class=None, *, lib=None, _stacklevel=1):
```
will refactor this following impl_atract

zou3519 replied:
We might want to think about what the best API for this is. Our options are:
- impl_abstract_class, like impl_abstract
- a method directly on the torchbind object.

Also, we really want two things:
1. a way to get the "state/metadata" of a TorchBind class
2. a way to convert the "state/metadata" into a Fake version of the TorchBind class.

I don't see those two things here (we want (1) by itself so that we can use it for guarding)

ydwu4 replied:
> a method directly on the torchbind object.

Are your referring to something like:
```python
class FakeFoo
  ...
foo_obj = torch.classes._Testing.Foo()
foo_obj.impl_fake(FakeFoo) # Or foo_obj.impl_fake(FakeFoo.from_real(foo_obj))?  
```
What's bad for `impl_abstrct_class`? It seems consistent with existing APIs. Does second approach bring additional benefits? My understanding is that it creates a direct relation between the TorchScript object and its corresponding fake implementation and perhaps less verbose compared with `impl_abstract_class`.

> a way to get state and convert to fake with meta data.

I'll add them shortly.

> create in a private namespace

Sounds good!

zou3519 replied:
> I'll add them shortly. After we have the get_metadata and fake_from_metadata, we no longer need from_real, right?

Yup 

> Are your referring to something like:

Something like:
```py
foo_obj = torch.classes._Testing.Foo()
metadata = foo_obj.__get_metadata__() 
fake_foo = foo_obj.__fake_from_metadata__(metadata)
```

> What's bad for impl_abstrct_class? It seems consistent with existing APIs. Does second approach bring additional benefits?

For tensor subclasses we defined `__tensor_flatten__`/`__tensor_unflatten__` as the way to retrieve and work with internal state. For operators we have impl_abstract. It's a question of which approach is "more consistent"

ydwu4 replied:
The interface looks good to me! Yeah, the consistency with the tensors are indeed better analogy than classes. Let me get it working.

zou3519 replied:
It's unclear to me if `__get_metadata__/__fake_from_metadata__` are better, because torchbind objects get defined in C++...

ydwu4 replied:
Yeah. To summarize: 
1. We need `__get_metadata__` anyway in dynamo guards so users have to implement this method. 
2. Users would also have to write a fake class in Python. 
3. When we encounter a script_object args for a custom op, we need a way to map the script_object to its fake class. The most straightforward way would be using the impl_atract_class apporach. Therefore users need to register this mapping.
4. users need to tell us how to create the fake class. Probably, we could ask users to implement fake_class.__fake_from_metadata__, where users must access the metadata via `__get_metadate__`. 

The end user-journey looks like this:
1. User pt2 compiles a model with torchScriptObject and sees an error message like ` cannot fakify the scriptObject and please register a fake class that implements a static method from_metadata`.
2. To write this fake class, users go back to C++ implementation and tries to collect the metadata required and add a __get_metadate__ methods with def_get_metadata. 
3. Then user implements the fake class and implements from_metadata with __get_metadata__.
4. We call the from_metadata call and continue tracing.

ydwu4 replied:
For dynamic shapes, user could call mark_dynamic in FakeFoo.from_real (or from_metadata).

ydwu4 committed:
Update on "Add torch.library.impl_abstract_class to fakify torchBind class"




[ghstack-poisoned]

aaronenyeshi commented:
@davidberard98, can you please help pull in the right TorchScript reviewers?

albanD commented on torch/autograd/profiler.py:
```diff
@@ -602,12 +602,17 @@ def __init__(self, name: str, args: Optional[str] = None):
-        )
+        from torch.fx.experimental.proxy_tensor import maybe_disable_fake_tensor_mode
+
+        with torch.utils._python_dispatch._disable_current_modes(), maybe_disable_fake_tensor_mode():
```
From my experience in touching this code a long time ago, doing anything but the bare minimum to call a single c++ op here is degrading performance significantly (specific use case for Meta employees is [here](https://fb.workplace.com/groups/1069285536500339/permalink/5027037044058482/))

ydwu4 replied:
Should we just special case this operator then?

zou3519 replied:
@ydwu4 why did we need additional handling around this?

ydwu4 replied:
`_record_function_enter_new, _record_function_exit._RecordFunction` defined one implementation for all [backends](https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/record_function_ops.cpp#L129-L140). 

[dynamo_timed](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/utils.py#L226-L273) used them in inductor to profile the compilation time and it's on the critical path. See the [test failure in previous version](https://hud.pytorch.org/pytorch/pytorch/pull/120045?sha=caa2d677aa22c37591f5d9d9edbaaf797314292b) if we don't fakify the inputs . 

If we fakify the input to _RecordFunction, the profiling is not executed. I believe that these profiler methods can be used to profile fake tensor propagation performance, so I decided to keep their behavior unchanged. However, adding maybe_disable_fake_tensor_mode before calling the op might incur too much overhead, according to Alban. Therefore, I came up with an additional handling approach.

For custom ops that have one implementation for all backends, should we fakify the TorchScriptObject inputs if running under fake tensor mode? My current thinking is that we should only fakify those inputs that users want to trace as a node in the graph module. For the rest, we should keep their original behavior since these operators indeed want to execute the real operation under fake mode. What do you think?

ydwu4 committed:
Update on "Add torch.library.impl_abstract_class to fakify torchBind class"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch.library.impl_abstract_class to fakify torchBind class"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch.library.impl_abstract_class to fakify torchBind class"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

[ghstack-poisoned]

ydwu4 commented:
@ydwu4 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54136016).

ydwu4 committed:
Update on "Add torch.library.impl_abstract_class to fakify torchBind class"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D54136016](https://our.internmc.facebook.com/intern/diff/D54136016)

[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch.library.impl_abstract_class to fakify torchBind class"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D54136016](https://our.internmc.facebook.com/intern/diff/D54136016)

[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch.library.impl_abstract_class to fakify torchBind class"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D54136016](https://our.internmc.facebook.com/intern/diff/D54136016)

[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch.library.impl_abstract_class to fakify torchBind class"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D54136016](https://our.internmc.facebook.com/intern/diff/D54136016)

[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch.library.impl_abstract_class to fakify torchBind class"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D54136016](https://our.internmc.facebook.com/intern/diff/D54136016)

[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch.library.impl_abstract_class to fakify torchBind class"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D54136016](https://our.internmc.facebook.com/intern/diff/D54136016)

[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch.library.impl_abstract_class to fakify torchBind class"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D54136016](https://our.internmc.facebook.com/intern/diff/D54136016)

[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch.library.impl_abstract_class to fakify torchBind class"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D54136016](https://our.internmc.facebook.com/intern/diff/D54136016)

[ghstack-poisoned]

zou3519 commented on torch/_library/abstract_impl_class.py:
```diff
@@ -0,0 +1,153 @@
+    """
+
+    def inner(fake_class):
+        global global_abstract_class_registry
```
Don't think this is needed, `global` is  only necessary if you re-assign the variable.

zou3519 commented on test/cpp/jit/test_custom_class_registrations.cpp:
```diff
@@ -339,9 +339,10 @@ TORCH_LIBRARY(_TorchScriptTesting, m) {
-  m.def(
-      "takes_foo(__torch__.torch.classes._TorchScriptTesting._Foo foo, Tensor x) -> Tensor");
+          })
+      .def_meta([](c10::intrusive_ptr<Foo> self) { // __get_metadata__
```
name bikeshedding:

Btw, if we're going for `__get_metadata__`, we probably want a better name for it. For Tensor Subclasses, the analogous API is named `__tensor_flatten__`.

ydwu4 replied:
`__obj_flatten__` then?

zou3519 commented on test/cpp/jit/test_custom_class_registrations.cpp:
```diff
@@ -339,9 +339,10 @@ TORCH_LIBRARY(_TorchScriptTesting, m) {
-  m.def(
-      "takes_foo(__torch__.torch.classes._TorchScriptTesting._Foo foo, Tensor x) -> Tensor");
+          })
+      .def_meta([](c10::intrusive_ptr<Foo> self) { // __get_metadata__
```
nb: we could totally just ask the user to `.def("__get_metadata__")` right now, right?

ydwu4 replied:
yeah, they can!

zou3519 commented on torch/testing/_internal/torchbind_impls.py:
```diff
@@ -0,0 +1,14 @@
+    torch.ops.load_library(str(lib_file_path))
+
+
+@torch.library.impl_abstract("_TorchScriptTesting::takes_foo")
+def fake_takes_foo(foo, z):
+    return foo.add_tensor(z)
```
Do we test this?

ydwu4 replied:
Yeah, this file is imported test_torchbind.py and when we do non-strict export for test_unlift_custom_obj.

ydwu4 committed:
Update on "Add torch.library.impl_abstract_class to fakify torchBind class"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D54136016](https://our.internmc.facebook.com/intern/diff/D54136016)

[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch.library.impl_abstract_class to fakify torchBind class"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D54136016](https://our.internmc.facebook.com/intern/diff/D54136016)

[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch.library.impl_abstract_class to fakify torchBind class"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D54136016](https://our.internmc.facebook.com/intern/diff/D54136016)

[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch.library.impl_abstract_class to fakify torchBind class"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D54136016](https://our.internmc.facebook.com/intern/diff/D54136016)

[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch.library.impl_abstract_class to fakify torchBind class"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D54136016](https://our.internmc.facebook.com/intern/diff/D54136016)

[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch.library.impl_abstract_class to fakify torchBind class"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D54136016](https://our.internmc.facebook.com/intern/diff/D54136016)

[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch._library.impl_abstract_class to fakify torchBind class"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D54136016](https://our.internmc.facebook.com/intern/diff/D54136016)

[ghstack-poisoned]

ydwu4 committed:
Update on "Add torch._library.impl_abstract_class to fakify torchBind class"




cc EikanWang jgong5 wenzhe-nrv sanchitintel

Differential Revision: [D54136016](https://our.internmc.facebook.com/intern/diff/D54136016)

[ghstack-poisoned]

zou3519 commented on torch/_library/abstract_impl_class.py:
```diff
@@ -0,0 +1,152 @@
+        TORCH_LIBRARY(_TorchScriptTesting, m) {
+            m.class_<Foo>("_Foo")
+                .def(torch::init<int64_t, int64_t>())
+                // .def(torch::init<>())
```
why's this commented out?

ydwu4 replied:
Oops, it's an error

ydwu4 commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -562,6 +562,11 @@ def wrapped(*proxies):
             lambda t: get_proxy_slot(t, tracer, t, lambda x: x.proxy),
             out
         )
+        out = pytree.tree_map_only(
+            torch.ScriptObject,
+            lambda t: get_proxy_slot(t, tracer, t, lambda x: x),
+            out
+        )
```
The output is a plain script object, we've tracked the script object at input and assoicated it with a proxy. This is just to retrieve its corresponding proxy. So that the `create_arg` of output node is able to connect its arguments with the correct proxy instead of creating a new one.

zou3519 commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -802,6 +802,45 @@ class DispatchCacheInfo:
+
+
+def _fakify_script_object(x):
+    if hash(x) in cached_fake_script_obj:
+        return cached_fake_script_obj[hash(x)]
```
Why do we need the caching? Do you have a concrete example?

zou3519 replied:
tq should already be a FakeTensorQueue when tracing -- Line 1 and Line 2 should not need to create new FakeTensorQueue. What am I missing here?

zou3519 commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1232,6 +1271,14 @@ def dispatch(self, func, types, args=(), kwargs=None):
+            torch.ops.profiler._record_function_enter_new,
+            torch.ops.profiler._record_function_exit._RecordFunction,
+        }:
+            args, kwargs = _maybe_fakify_script_object(args, kwargs)
```
How did you run into a case where we are fakifying inputs to torch.ops.profiler?

ydwu4 replied:
_record_function_enter_new, _record_function_exit._RecordFunction defined one implementation for all [backends](https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/record_function_ops.cpp#L129-L140).

[dynamo_timed](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/utils.py#L226-L273) used them in inductor to profile the compilation time and it's on the critical path. [Test failure of triggering the fakification](https://hud.pytorch.org/pytorch/pytorch/pull/120045?sha=caa2d677aa22c37591f5d9d9edbaaf797314292b) 
```
2024-02-16T01:36:43.4777090Z     torch.ScriptObject, lambda x: _wrap_script_object(x), (args, kwargs)
2024-02-16T01:36:43.4778229Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_subclasses/fake_tensor.py", line 1200, in _wrap_script_object
2024-02-16T01:36:43.4779290Z     fake_class = torch.library.registered_class[x._type().qualified_name()]
2024-02-16T01:36:43.4780112Z torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
2024-02-16T01:36:43.4780857Z KeyError: '__torch__.torch.classes.profiler._RecordFunction'
```

zou3519 commented on test/export/test_torchbind.py:
```diff
@@ -1,36 +1,83 @@
+
+            @classmethod
+            def from_real(cls, real_tq):
+                return cls(real_tq.clone_queue())
```
1. clone_queue() is not good enough because the strides on the cloned queue may be different from the strides on the original queue. The way to go about it is to use FakeTensorConverter to convert each element of the queue.

zou3519 replied:
We should probably add an explicit test for this (that takes a TensorQueue and creates a FakeTensorQueue and then checks the properties of the Tensors)

ydwu4 replied:
Yeah, but this from_real is written by users. We won't want them to use FakeTensorConverter directly... Let me figure out how to deal with it. Right now, the tensors inside the object is not fakified yet.

zou3519 replied:
Does directly using the FakeTensor constructor hit FakeTensorConverter?

ydwu4 replied:
no. FakeTensorMode.from_real hits the converter.

zou3519 commented on test/export/test_torchbind.py:
```diff
@@ -1,36 +1,83 @@
+                continue
+            self.assertEqual(a, b)
+
+    def _test_export_same_as_eager_skip_script_object(
```
We have a lot of ways of naming the same thing. We should try to be consistent when we're creating new APIs. I've seen "script object", "custom class", "torchbind class", etc...

We should probably just call it TorchBind{Object/Class}

ydwu4 replied:
the class name is torch.ScriptObject though..hahah But never mind. TorchBindObject/Class seems good to me.

zou3519 replied:
So we do have the option of changing torch.ScriptObject into torch.TorchBindObject. But torch.TorchBindObject sounds a bit wild... should we call it "CustomClass/CustomObject" then?

zou3519 commented on test/export/test_torchbind.py:
```diff
@@ -1,36 +1,83 @@
+        flat_actual = pytree.tree_leaves(actual)
+        self.assertEqual(len(flat_exp), len(flat_actual))
+        for a, b in zip(flat_exp, flat_actual):
+            if isinstance(a, torch.ScriptObject) and isinstance(b, torch.ScriptObject):
+                continue
```
well you could define a `__eq__` on the ScriptObject if you wanted to :P
It might be good to do to check the state of the ScriptObjects are the same?

ydwu4 replied:
This `__eq__` varies from object to object, right? It's probably difficult to migrate one for all existing ones...

zou3519 replied:
fair enough

zou3519 commented on torch/export/graph_signature.py:
```diff
@@ -85,7 +85,10 @@ class OutputSpec:
-        assert isinstance(self.arg, (TensorArgument, SymIntArgument, ConstantArgument))
+        assert isinstance(
+            self.arg,
+            (TensorArgument, SymIntArgument, ConstantArgument, CustomObjArgument),
```
What's a `CustomObjArgument`? Is this more general than torch.ScriptObject?

ydwu4 replied:
This is exactly torch.ScriptObject, and we probably should re-name it that way.

ydwu4 commented on torch/_library/abstract_impl_class.py:
```diff
@@ -0,0 +1,152 @@
+
+            def add_tensor(self, z):
+                return (self.x + self.y) * z1
+
```
Will add it.

zou3519 commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1232,6 +1271,14 @@ def dispatch(self, func, types, args=(), kwargs=None):
+            torch.ops.profiler._record_function_enter_new,
+            torch.ops.profiler._record_function_exit._RecordFunction,
+        }:
+            args, kwargs = _maybe_fakify_script_object(args, kwargs)
```
This should not be where we're Fake-ifying the ScriptObject. For example, FakeTensorMode does not allow non-fake-Tensors by default (https://github.com/pytorch/pytorch/blob/c78f72d7e7965f8dc9516f462c2c78e140d4e57c/torch/_subclasses/fake_tensor.py#L824).

When we're doing make_fx tracing, make_fx should be fakeifying any ScriptObject inputs

ydwu4 replied:
I tried that approach: i.e. fakifying the inputs like wrap. But this causes problem downstreams when cooperating with custom ops, because they expect the script object to be the real one in the schema.

zou3519 replied:
Do you know if the custom op expects a "TensorQueue" or if it expects a `torch::CustomClassHolder`?

zou3519 replied:
Using the FakeTensor-Tensor relationship as an analogy:
- FakeTensorQueue should be passable into C++.
- One should be able to call FakeTensorQueue methods in C++.
- Open question on what happens if TensorQueue has any public non-method members...

zou3519 replied:
I guess we need to ban public non-method members on these custom classes. Otherwise, there's really not going to be a way for us to write a "Fake" version of the class.

ydwu4 replied:
> Do you know if the custom op expects a "TensorQueue" or if it expects a torch::CustomClassHolder?

It's expecting a TensorQueue.

> FakeTensorQueue probably needs to be a subclass (somehow) of TensorQueue ?

It seems that if we're going in this way, we are re-inventing tensor subclasses for torchbind classes? In the end probably requires extending the dispatcher infra as well?

zou3519 commented on test/export/test_torchbind.py:
```diff
@@ -1,36 +1,83 @@
+                self.y = y
+
+            @classmethod
+            def from_real(cls, foo):
+                (x, y), _ = foo.__getstate__()
+                return cls(x, y)
```
nit: how do you feel about renaming `from_real` to `from_concrete`? "concrete/abstract" seem like opposites, while "real/fake" are opposites (and we did not name our API "impl_fake").

zou3519 reviewed (commented):
this mostly looks good to me. high-level comments:
- I don't think we're fakeifying the ScriptObject in the correct place. FakeTensor's `__torch_dispatch__` should not Fakeify every ScriptObject it sees (instead it should error out on it). Instead, make_fx(f, tracing_mode=fake)(*args) should fakeify args.
- We should add a loud warning on the mutation problem. Mutating an object that gets put inside a TorchBind object will lead to silent incorrectness, unless we figure out some way to surface it to the user.
